test/test_torch.py

```diff
@@ -5902,80 +5908,67 @@ def run(device, data, model, optimizer, scaler, loss_fn, skip_iter, try_scaling_
+            device, dtype, optim_info, skip=("differentiable"))
+
+        # Compares no scaling + no autocasting against scaling + autocasting.
+        for optim_input in all_optim_inputs:
+
+            # NOTE(mkozuki): With current way of testing, `torch.optim.Adam` is failing in spite of `foreach` and `fused`.
+            #   Giving some flexibility to this test might help.
+            context = contextlib.nullcontext
+            if optimizer_ctor in (torch.optim.Adam, torch.optim.AdamW):
+                from functools import partial
+                context = partial(self.assertRaises, AssertionError)
+            with context():
+                # sets atol=1e-3 because we're comparing pure fp32 arithmetic vs a mixture of fp16 and fp32
+                self._run_scaling_case(
+                    device, run, unskipped=3, skipped=1, atol=1e-3,
+                    optimizer_ctor=optimizer_ctor, optimizer_kwargs=optim_input.kwargs,
+                )
+                # this will be picked up by try_pickle within run():
+                try_pickle = True
+                self._run_scaling_case(
+                    device, run, unskipped=3, skipped=1, atol=1e-3,
+                    optimizer_ctor=optimizer_ctor, optimizer_kwargs=optim_input.kwargs,
+                )
 
     # Make sure that the parameters become nonsense when scaled gradients are finite
     # but they get invalidated before `optimizer.step`, after `GradScaler.unscale_`
 
     @onlyNativeDeviceTypes
-    def test_params_invalidated_with_grads_invalidated_between_unscale_and_step(self, device):
-        device = torch.device(device)
-        for optimizer_ctor, optimizer_kwargs in product(
-            (torch.optim.Adam, torch.optim.AdamW),
-            (
-                {"foreach": False, "fused": False},
-                {"foreach": True, "fused": False},
-                {"foreach": False, "fused": True},
-            ),
-        ):
-            with self.subTest(optimizer=optimizer_ctor, optimizer_kwargs=optimizer_kwargs):
-                self._test_grads_invalidated_between_unscale_and_step(device.type, optimizer_ctor, optimizer_kwargs)
-
-    def _test_grads_invalidated_between_unscale_and_step(self, device, optimizer_ctor, optimizer_kwargs):
-        model, _, optimizer, _, data, loss_fn, _ = _create_scaling_case(
-            device, optimizer_ctor=optimizer_ctor, optimizer_kwargs=optimizer_kwargs,
-        )
-        scaler = torch.GradScaler(device=device, init_scale=128.0)
+    @optims(
+        [optim for optim in optim_db if optim.optim_cls in [torch.optim.AdamW, torch.optim.Adam, torch.optim.SGD]],
+        dtypes=[torch.float32]
+    )
+    def test_params_invalidated_with_grads_invalidated_between_unscale_and_step(self, device, dtype, optim_info):
+        optimizer_ctor = optim_info.optim_cls
+        all_optim_inputs = _get_optim_inputs_including_global_cliquey_kwargs(
+            device, dtype, optim_info, skip=("differentiable"))
```

Comment:
```suggestion
            device, dtype, optim_info, skip=("differentiable",))
```
make it a tuple! right now it’s just a string